src/org/opensolaris/opengrok/configuration/messages/handler/ProjectMessageHandler.java

```diff
@@ -0,0 +1,330 @@
+                break;
+            case "list":
+                return Response.of(env.getProjectNames().stream().collect(Collectors.joining("\n")));
+            case "list-indexed":
+                return Response.of(env.getProjectList().stream().filter(Project::isIndexed).
+                        map(Project::getName).collect(Collectors.joining("\n")));
+            case "get-repos":
+                return Response.of(mapRepositories(message.getTags(), RepositoryInfo::getDirectoryNameRelative)
+                        .collect(Collectors.joining("\n")));
+            case "get-repos-type":
+                return Response.of(mapRepositories(message.getTags(), RepositoryInfo::getType).distinct().sorted()
+                        .collect(Collectors.joining("\n")));
+            default:
+                throw new HandleException("Unknown command " + command);
+        }
+
+        return Response.of("command \"" + message.getText() + "\" for projects " +
+                String.join(",", message.getTags()) + " done");
+    }
+
+    /**
+     * Perform additional validation. This cannot be done in validate()
+     * because it does not have access to the currently used RuntimeEnvironment.
+     */
+    private void validateMore(RuntimeEnvironment env, Message message) throws ValidationException {
+        String command = message.getText();
+
+        if (!env.isProjectsEnabled()) {
+            throw new ValidationException("Projects have to be enabled in order to use this message");
+        }
+
+        // The same check can be done for the "delete" command however it is better
+        // to allow for some flexibility. The source can be deleted first and then
+        // the "delete" command can follow.
+        if (command.equals("add")) {
+            File srcRoot = env.getSourceRootFile();
+            for (String projectName : message.getTags()) {
+                File projDir = new File(srcRoot, projectName);
+                if (!projDir.isDirectory()) {
+                    throw new ValidationException(projDir.getAbsolutePath() + " is not a directory");
+                }
+            }
+        }
+
+        if (command.equals("delete")) {
+            for (String projectName : message.getTags()) {
+                if (!env.getProjects().containsKey(projectName)) {
+                    throw new ValidationException("project \"" + projectName +
+                            "\" not found in configuration");
+                }
+            }
+        }
+    }
+
+    private void addProjects(final Set<String> projectNames) {
+        for (String projectName : projectNames) {
+            File srcRoot = env.getSourceRootFile();
+            File projDir = new File(srcRoot, projectName);
+
+            if (!env.getProjects().containsKey(projectName)) {
+                Project project = new Project(projectName, "/" + projectName);
+                project.setTabSize(env.getConfiguration().getTabSize());
```

Comment:
what about `project.completeWithDefaults(env.getConfiguration());`

Reply:
Again, copied code, I will look into it.